### ISSUE TITLE
Training data collection

### ISSUE THREAD

**Issue by SwayamDani** [SwayamDani/special-eureka#5]
# Copilot Prompt for Political LLM Training Data Collection

## Task Overview
Create a comprehensive data collection system to gather free online political content for fine-tuning a Llama 3.1 8B model that generates professional political campaign strategies. The system should collect 2500+ high-quality examples in instruction-following format.

## Required Output Format
All collected data must be formatted as:
```json
{
  "instruction": "Create a specific political strategy task",
  "input": "Context: Party name, election type, state, current political situation",
  "response": "Detailed professional strategy with budgets, timelines, KPIs, and implementation steps (800-2000 words)"
}
```

### 1. Political Analysis Articles (1000 examples)
**Free Sources to Scrape:**
```python
political_news_sources = [
    "theprint.in",
    "scroll.in", 
    "thewire.in",
    "newsclick.in",
    "caravanmagazine.in",
    "sabrangindia.in",
    "forwardpress.in",
    "roundtableindia.co.in"
]
```

**What to Extract:**
- Electoral analysis and campaign strategies
- Alliance formation discussions
- Regional political dynamics
- Opposition research and counter-strategies
- Campaign post-mortems and success stories

**Scraping Requirements:**
- Target articles from 2020-2024 for current relevance
- Extract full article text, headlines, and metadata
- Filter for articles >500 words with political strategy content
- Remove duplicates and low-quality content

### 2. Regional Language Political Content (500 examples)
**Hindi Sources:**
```python
hindi_sources = [
    "livehindustan.com",
    "jagran.com", 
    "amarujala.com",
    "patrika.com",
    "dainikbhaskar.com"
]
```

**Regional Sources:**
```python
regional_sources = {
    "Tamil": ["dinamalar.com", "thanthi.com"],
    "Telugu": ["eenadu.net", "sakshi.com"],
    "Bengali": ["anandabazar.com", "bartamanpatrika.com"],
    "Malayalam": ["mathrubhumi.com", "madhyamam.com"],
    "Marathi": ["loksatta.com", "sakal.com"]
}
```

### 3. Academic Research Papers (200 examples)
**Free Academic Sources:**
```python
academic_sources = [
    "shodhganga.inflibnet.ac.in",  # Indian PhD theses
    "archive.org",                 # Free books and papers
    "jstor.org",                   # Free articles
    "cpsindia.org",               # Policy research
    "orfonline.org",              # Strategic studies
    "idsa.in",                    # Defence studies
    "vifindia.org"                # International studies
]
```

**Focus Areas:**
- Indian electoral studies and campaign analysis
- Regional political dynamics and caste politics
- Alliance strategies and coalition building
- Digital campaigning and social media strategies

### 4. Government & Election Data (300 examples)
**Official Sources:**
```python
government_sources = [
    "eci.gov.in",                 # Election Commission reports
    "myneta.info",                # Electoral data
    "adrindia.org",               # Data analysis
    "prsindia.org",               # Legislative research
    "parliamentofindia.nic.in"    # Parliamentary proceedings
]
```

### 5. Opinion & Editorial Content (500 examples)
**Free Editorial Sources:**
- The Hindu (free articles)
- Indian Express (free content)
- Deccan Herald political analysis
- Outlook India political features
- Frontline magazine archive

### Web Scraping Framework
```python
# Required libraries and approach
import requests
from bs4 import BeautifulSoup
import pandas as pd
import json
from newspaper import Article
import time
from selenium import webdriver

# Scraping guidelines
scraping_rules = {
    "delay_between_requests": 2,  # Ethical scraping
    "user_agent": "Mozilla/5.0...", # Proper headers
    "retry_attempts": 3,
    "timeout": 10,
    "content_filter": "min_length > 500 words"
}
```

### Data Processing Pipeline
```python
def process_political_content(raw_content):
    # 1. Extract and clean text
    # 2. Identify political keywords and entities
    # 3. Categorize by political topic
    # 4. Format for instruction-following
    # 5. Quality validation
    return formatted_training_data
```

### Content Enhancement System
Create a system to convert raw articles into instruction-following format:

**Example Transformation:**
```
Raw Article: "Analysis of BJP's 2024 campaign strategy in UP shows focus on..."
↓
Instruction: "Analyze BJP's campaign strategy for 2024 UP elections"
Input: "Recent electoral trends show BJP facing challenges in rural areas"
Response: "[Detailed strategy analysis based on article content with specific recommendations]"
```

### Content Filtering Criteria
```python
quality_filters = {
    "minimum_length": 500,
    "political_keyword_density": 0.05,
    "publication_date": "2020-2024",
    "language_detection": "accurate",
    "duplicate_threshold": 0.8,
    "relevance_score": 0.7
}
```

### Political Keyword Lists
```python
political_keywords = [
    # Strategy terms
    "campaign", "alliance", "coalition", "manifesto", "strategy",
    "electoral", "voter", "constituency", "candidate", "party",
    
    # Indian political terms  
    "lok sabha", "vidhan sabha", "bjp", "congress", "aap",
    "caste", "community", "dalit", "obc", "minority",
    
    # Tactical terms
    "booth", "mobilization", "outreach", "messaging", "digital"
]
```

### Training Dataset Format
```json
{
  "metadata": {
    "source": "theprint.in",
    "date": "2024-01-15",
    "language": "english",
    "category": "electoral_strategy",
    "state": "uttar_pradesh",
    "quality_score": 0.85
  },
  "training_example": {
    "instruction": "Develop alliance strategy for regional party",
    "input": "Party: Samajwadi Party, Election: 2027 Assembly, State: UP, Context: Recent electoral setbacks",
    "response": "[Professional strategy with specific action items, budgets, timelines]"
  }
}
```

### File Organization
```
political_training_data/
├── english_analysis/           # 1000 examples
├── hindi_content/             # 300 examples  
├── regional_languages/        # 200 examples
├── academic_papers/           # 200 examples
├── government_reports/        # 300 examples
├── editorials_opinions/       # 500 examples
└── processed_dataset.json     # Final training file
```

### Progress Tracking
Create a system to track:
- Articles collected per source
- Processing success rates
- Quality score distributions
- Language/region coverage
- Training data completeness

### Data Quality Checks
1. **Content Relevance**: Political strategy focus
2. **Language Quality**: Grammar and coherence
3. **Factual Accuracy**: Cross-reference with known events
4. **Instruction Clarity**: Clear task formulation
5. **Response Quality**: Professional consultant-grade output

### Sample Testing
Before full collection, test the pipeline with 50 examples from each source to validate:
- Scraping accuracy
- Content extraction quality  
- Instruction generation effectiveness
- Response formatting consistency

### Final Dataset Composition
- **Total Examples**: 2500+
- **Languages**: English (70%), Hindi (20%), Regional (10%)
- **Time Coverage**: 2020-2024 (80% post-2022)
- **Geographic Coverage**: All major Indian states
- **Quality Score**: >0.7 average

### Delivery Format
```
1. Raw collected data (JSON files by source)
2. Processed training dataset (Hugging Face format)
3. Quality analysis report
4. Collection statistics and metrics
5. Documentation and usage instructions
```

## Implementation Timeline

**Week 1**: Setup scraping infrastructure, test on 100 examples
**Week 2**: Scale to 500 examples, implement quality filters  
**Week 3**: Reach 1500 examples, add regional languages
**Week 4**: Complete 2500+ examples, final processing
**Week 5**: Quality validation, formatting, documentation
**Week 6**: Delivery and testing integration

## Success Metrics

- **Collection Rate**: >2500 high-quality examples
- **Quality Score**: Average >0.75
- **Coverage**: All major political parties and states
- **Freshness**: 80% content from 2022-2024
- **Diversity**: Balanced across strategy types and regions

Create a comprehensive, automated system that can continuously collect and process this political content while maintaining high quality standards and ethical scraping practices.
